promise-bot/fix_promise_items.py

```diff
@@ -0,0 +1,257 @@
+"""
+Fix promise items
+"""
+import argparse
+import errno
+import sys
+
+from configparser import ConfigParser
+from pathlib import Path
+
+from olclient.openlibrary import OpenLibrary, Config
+
+DEFAULT_BATCH_SIZE = 1000
+DEFAULT_START_LINE = 1
+IN_FILE = "./input/in.txt"
+STATE_FILE = "./input/curline.txt"
+ERR_FILE = "./out/err.txt"
+
+
+class FixPromiseItems:
+    def __init__(
+        self,
+        in_file,
+        state_file,
+        error_file,
+        ol=None,
+        batch_size=DEFAULT_BATCH_SIZE,
+        start_line=DEFAULT_START_LINE,
+        dry_run=True,
+    ):
+        self.ol = ol
+
+        self.modified = 0
+        self.errors = 0
+        self.matched = 0
+
+        self.in_file = in_file
+        self.error_file = error_file
+        self.state_file = state_file
+
+        self.batch_size = batch_size
+        self.start_line = start_line
+        self.dry_run = dry_run
+
+    def run(self):
+        with open(self.in_file, "r") as f:
+            if self.start_line:
+                for _ in range(1, self.start_line):
+                    f.readline()
+            for _ in range(self.batch_size):
+                line = f.readline()
+                if not line:
+                    break
+                edition_olid = self.extract_olid(line[:-1])  # Trim trailing newline
+
+                try:
+                    edition = self.ol.get(edition_olid)
+                    self.update_edition(edition)
+                    if not self.dry_run:
+                        edition.save("Fix bwbsku, bwb_id, amz_id, source_record")
+                    self.modified += 1
+                except Exception:
+                    self.write_error(self.error_file, line[:-1])
+                    self.errors += 1
+                    pass
+
+        num_processed = self.modified + self.errors + self.matched
+        self.write_state(self.state_file, self.start_line + num_processed)
+
+        return {
+          'processed': num_processed,
+          'modified': self.modified,
+          'matched': self.matched,
+          'errors': self.errors,
+        }
+    def extract_olid(self, line):
+        fields = line.split("\t")
+        return fields[1].split("/")[-1]
+
+    def update_edition(self, edition):
+        id_count = len(edition.local_id)
+
+        local_id = next(
+            (b for b in edition.local_id if b.startswith("urn:bwbsku:")), ""
+        )
+
+        # Fix local IDs:
+        updated_ids = [local_id] + [
+            _id for _id in edition.local_id if not _id.startswith("urn:bwbsku:")
+        ]
+
+        # Check if record was previously modified:
+        if id_count == len(updated_ids):
+            self.matched += 1
+            return
+
+        edition.local_id = updated_ids
+
+        # Fix identifiers:
+        edition.identifiers.pop("amazon", None)
+        edition.identifiers.pop("better_world_books", None)
+
+        # Fix source_records entry:
+        sku = local_id.split(":")[-1]
```

Comment:
If `local_id` is `""` are we going to get `source_ids` that look like `{source_record}:`?

Maybe fine

Reply:
Throwing an exception and not writing to catalog if sku is empty.